core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushJoinIntoTableScan.java

```diff
@@ -68,131 +75,145 @@ public class PushJoinIntoTableScan
+    private final Pattern<JoinNode> pattern;
 
-    public PushJoinIntoTableScan(PlannerContext plannerContext)
+    public static Set<Rule<?>> rules(PlannerContext plannerContext)
     {
-        this.plannerContext = requireNonNull(plannerContext, "plannerContext is null");
+        return ImmutableSet.of(
+                new PushJoinIntoTableScan(plannerContext, ProjectSide.LEFT),
+                new PushJoinIntoTableScan(plannerContext, ProjectSide.RIGHT),
+                new PushJoinIntoTableScan(plannerContext, ProjectSide.BOTH),
+                new PushJoinIntoTableScan(plannerContext, ProjectSide.NONE));
     }
 
-    @Override
-    public Pattern<JoinNode> getPattern()
+    PushJoinIntoTableScan(PlannerContext plannerContext, ProjectSide projectSide)
     {
-        return PATTERN;
+        this.projectSide = requireNonNull(projectSide, "projectSide is null");
+        this.plannerContext = requireNonNull(plannerContext, "plannerContext is null");
+        this.pattern = switch (projectSide) {
+            case LEFT -> Patterns.join()
+                    .with(left().matching(project().capturedAs(LEFT_PROJECT).with(source().matching(tableScan().capturedAs(LEFT_TABLE_SCAN)))))
+                    .with(right().matching(tableScan().capturedAs(RIGHT_TABLE_SCAN)));
+            case RIGHT -> Patterns.join()
+                    .with(left().matching(tableScan().capturedAs(LEFT_TABLE_SCAN)))
+                    .with(right().matching(project().capturedAs(RIGHT_PROJECT).with(source().matching(tableScan().capturedAs(RIGHT_TABLE_SCAN)))));
+            case BOTH -> Patterns.join()
+                    .with(left().matching(project().capturedAs(LEFT_PROJECT).with(source().matching(tableScan().capturedAs(LEFT_TABLE_SCAN)))))
+                    .with(right().matching(project().capturedAs(RIGHT_PROJECT).with(source().matching(tableScan().capturedAs(RIGHT_TABLE_SCAN)))));
+            case NONE -> Patterns.join()
+                    .with(left().matching(tableScan().capturedAs(LEFT_TABLE_SCAN)))
+                    .with(right().matching(tableScan().capturedAs(RIGHT_TABLE_SCAN)));
+        };
     }
 
     @Override
-    public boolean isEnabled(Session session)
+    public Pattern<JoinNode> getPattern()
     {
-        return isAllowPushdownIntoConnectors(session);
+        return pattern;
     }
 
     @Override
     public Result apply(JoinNode joinNode, Captures captures, Context context)
     {
-        if (joinNode.isCrossJoin()) {
+        PlanNodes nodes = PlanNodes.create(captures, joinNode, projectSide);
+
+        if (nodes.join.isCrossJoin()) {
             return Result.empty();
         }
 
-        TableScanNode left = captures.get(LEFT_TABLE_SCAN);
-        TableScanNode right = captures.get(RIGHT_TABLE_SCAN);
-
-        verify(!left.isUpdateTarget() && !right.isUpdateTarget(), "Unexpected Join over for-update table scan");
+        if (nodes.leftScan.isUpdateTarget() || nodes.rightScan.isUpdateTarget()) {
```

Comment:
Use the public interface for the PlanNodes record (i.e., `.leftScan()` and `.rightScan()`). Same comment everywhere the fields of PlanNodes are accessed.